### PR TITLE
deflake more tests

### PR DESCRIPTION
**Pull request by i0rek**
This PR will print all goroutines when the tests panic and it also add support for golang 1.13.

On top of that there are some more fixed tests now:

* TestRaft_SnapshotRestore_PeerChange: This test was blocked on writing back the response into the channel from the RPC by a follower. Fixed by making it a buffered channel of size 1 and by changing the future to react to raft shutdown.
* TestRaft_UserRestore: before it would error when leadership changes because of the restore which is entirely possible though.

### CODE REVIEW

**Comment by i0rek** [comment]
Now others are failing. 😭

File: testing.go [hunk 1]
```diff
@@ -288,27 +289,29 @@ func (c *cluster) WaitEventChan(filter FilterFn, timeout time.Duration) <-chan s
 			select {
 			case <-timeoutCh:
 				return
-
+			case <-doneCh:
```
Comment (s-christoff) [inline testing.go]: Hey, could you ELI5 the difference between the `timeoutCh` and `doneCh`  🙏

Comment (i0rek) [reply testing.go]: `WaitEventChan` spawns a goroutine but there was no way to stop it. It only stopped before on timeout, but if we know that we don't need it anymore, it can now be stopped by closing doneCh. There are other ways of doing this, eg using context.

Comment (s-christoff) [reply testing.go]: Thank makes sense, tyvm <3

Comment (i0rek) [reply testing.go]: Switched to context.

File: .circleci/config.yml [hunk 2]
```diff
@@ -4,9 +4,11 @@ references:
   images:
     go-1.11: &GOLANG_1_11_IMAGE circleci/golang:1.11
     go-1.12: &GOLANG_1_12_IMAGE circleci/golang:1.12
+    go-1.13: &GOLANG_1_13_IMAGE circleci/golang:1.13
```
[the remark this answers is not in the document]
Comment (i0rek) [reply .circleci/config.yml]: 👍

File: testing.go [hunk 3]
```diff
@@ -276,8 +276,9 @@ func (c *cluster) Close() {
 // or a timeout occurs. It is possible to set a filter to look for specific
 // observations. Setting timeout to 0 means that it will wait forever until a
 // non-filtered observation is made.
-func (c *cluster) WaitEventChan(filter FilterFn, timeout time.Duration) <-chan struct{} {
+func (c *cluster) WaitEventChan(filter FilterFn, timeout time.Duration) (<-chan struct{}, chan struct{}) {
```
[the remark this answers is not in the document]
Comment (i0rek) [reply testing.go]: I am using context now.

File: raft_test.go [hunk 4]
```diff
@@ -1269,6 +1272,8 @@ func snapshotAndRestore(t *testing.T, offset uint64) {
 	}
 	fsm.Unlock()
 
+	// Make sure to get the possibly new leader
+	leader = c.Leader()
```
[the remark this answers is not in the document]
Comment (i0rek) [reply raft_test.go]: Well that is a good comment, thanks for reminding me. I will do some research around that to make sure this is actually correct. It seems intuitive to me, but it wouldn't be the first time I am wrong :D.

[the remark this answers is not in the document]
Comment (i0rek) [reply raft_test.go]: I think I was wrong! I double checked and what happens is that when the followers are restoring the snapshots, sometimes the heartbeats are timing out and the leader steps down:

```
    testing.go:150:  ... failed to contact quorum of nodes, stepping down
```

I increased the timeouts for our inmem raft config, lets see if that makes it any better.

[the remark this answers is not in the document]
Comment (i0rek) [reply raft_test.go]: It doesn't seem to help. I am putting back the old config and calling it a day for now. These tests are soo flaky :(.

**Comment by i0rek** [comment]
I took out all the controversial things, and it should be good to merge now.

File: raft_test.go [hunk 5]
```diff
@@ -208,7 +208,9 @@ func TestRaft_RecoverCluster(t *testing.T) {
 		c.EnsureSamePeers(t)
 	}
 	for applies := 0; applies < 20; applies++ {
-		runRecover(applies)
+		t.Run(fmt.Sprintf("%d applies", applies), func(t *testing.T) {
```
Comment (notnoop) [inline raft_test.go]: Love `t.Run()` use!
